src/llama.cpp

```diff
@@ -14777,12 +14979,19 @@ static std::vector<llama_vocab::id> llama_tokenize_internal(const llama_vocab &
                 // tokenizer.encode('', add_special_tokens=False) returns []
 
                 bool is_prev_special = false;
-
                 if (add_special && vocab.tokenizer_add_bos) {
                     GGML_ASSERT(vocab.special_bos_id != -1);
                     output.push_back(vocab.special_bos_id);
                     is_prev_special = true;
                 }
+                // add prefix to chatglm3
+                if (vocab.type_pre == LLAMA_VOCAB_PRE_TYPE_CHATGLM3) {
+                    output.push_back(64790); // [gMask]
+                    output.push_back(64792); // sop
+                    output.push_back(64795); // <|user|>
+                    output.push_back(30910); // \n
+                    output.push_back(13);
+                }
```

Comment:
These prefixes and suffixes should not by applied here. Instead, the user code is responsible for that (e.g. chat templates)

Reply:
@youth123 I didn't noticed this part, maybe this explain why the model's answer is incorrect: because the `[gMask]sop<|user|>` is added twice, once here and once in the chat template. We should remove it here.

Reply:
Incorrect answers have nothing to do with being added twice. I have compared the input of the hf model output before, and the output of each layer of the transformer is percentile aligned. But the error becomes larger in the last layer, and I'm still looking at this strange problem now. And regarding chat template, will it run in conversation mode in the future?